src/python/pants/backend/python/lint/pyupgrade/rules.py

```diff
@@ -39,17 +39,30 @@ class PyUpgradeRequest(FixTargetsRequest):
 async def pyupgrade_fix(request: PyUpgradeRequest.Batch, pyupgrade: PyUpgrade) -> FixResult:
     pyupgrade_pex = await Get(VenvPex, PexRequest, pyupgrade.to_pex_request())
 
-    result = await Get(
-        FallibleProcessResult,
-        VenvPexProcess(
-            pyupgrade_pex,
-            argv=(*pyupgrade.args, *request.files),
-            input_digest=request.snapshot.digest,
-            output_files=request.files,
-            description=f"Run pyupgrade on {pluralize(len(request.files), 'file')}.",
-            level=LogLevel.DEBUG,
-        ),
-    )
+    # NB: Pyupgrade isn't idempotent, but eventually converges. So keep running until it stops
+    # changing code. See https://github.com/asottile/pyupgrade/issues/703
+    # (Technically we could not do this. It doesn't break Pants since the next run on the CLI would
+    # use the new file with the new digest. However that isn't the UX we want for our users.)
+    num_changes_left = 10  # Give the loop an upper bound to guard against inifite runs
+    input_digest = request.snapshot.digest
+    while num_changes_left:
+        result = await Get(
+            FallibleProcessResult,
+            VenvPexProcess(
+                pyupgrade_pex,
+                argv=(*pyupgrade.args, *request.files),
+                input_digest=input_digest,
+                output_files=request.files,
+                description=f"Run pyupgrade on {pluralize(len(request.files), 'file')}.",
+                level=LogLevel.DEBUG,
+            ),
+        )
+        if input_digest == result.output_digest:
+            # Nothing changed, either due to failure or because it is fixed
+            break
+        input_digest = result.output_digest
+        num_changes_left -= 1
+
```

Comment:
If we hit num_changes_left == 0 should we error? Or at least warn? We're leaving the code in some unstable state.

Reply:
It's not unstable. It just might require re-running. I think we'll log an error.

Reply:
yea, this is a highly unlikely scenario, but if the state is broken that would mean that pyupgrade broke it.